### PR TITLE
Add POP kid to SuggestedCacheKey for AcquireTokenForClient

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes # TBD

**Changes proposed in this request**

**Testing**

**Performance impact**

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/ApiConfig/AbstractConfidentialClientAcquireTokenParameterBuilder.cs [hunk 1]
```diff
@@ -42,32 +42,6 @@ public override Task<AuthenticationResult> ExecuteAsync(CancellationToken cancel
...
-        /// <returns>The builder.</returns>
-        /// <remarks>
-        /// <list type="bullet">
-        /// <item><description>An Authentication header is automatically added to the request</description></item>
-        /// <item><description>The PoP token is bound to the HTTP request, more specifically to the HTTP method (GET, POST, etc.) and to the Uri (path and query, but not query parameters).</description></item>
-        /// <item><description>MSAL creates, reads and stores a key in memory that will be cycled every 8 hours.</description></item>
-        /// <item><description>This is an experimental API. The method signature may change in the future without involving a major version upgrade.</description></item>
-        /// </list>
-        /// </remarks>
-        public T WithProofOfPossession(PoPAuthenticationConfiguration popAuthenticationConfiguration)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AbstractConfidentialClientAcquireTokenParameterBuilder.cs:61]: I moved `WithProofOfPossession` to AcquireTokenForClient only. Would be a breaking change if not for experimental flag.

File: src/client/Microsoft.Identity.Client/Cache/CacheKeyFactory.cs [hunk 2]
```diff
@@ -76,18 +77,23 @@ private static bool GetOboOrAppKey(AuthenticationRequestParameters requestParame
...
             key = null;
             return false;
         }
 
-        public static string GetClientCredentialKey(string clientId, string tenantId)
-        {                
-            return $"{clientId}_{tenantId}_AppTokenCache";
+        public static string GetClientCredentialKey(string clientId, string tenantId, string popKid)
+        {
+            return string.Format(
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/CacheKeyFactory.cs:91]: string.Format invariant might a tad faster than $"" which is culture sensitive.

File: tests/Microsoft.Identity.Test.Unit/pop/PopAuthenticationSchemeTests.cs [hunk 3]
```diff
@@ -140,7 +139,7 @@ public async Task ValidateKeyExpirationAsync()
                 var initialToken = result.AccessToken;
 
                 //Advance time 7 hours. Should still be the same key and token
-                testClock.TestTime +=TimeSpan.FromHours(7);
+                testClock.UtcNow +=TimeSpan.FromHours(7);
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Unit/pop/PopAuthenticationSchemeTests.cs:142]: The name Utsnow on the test clock seems a bit misleading. Seems like it should always be the current time but it changes as modifications are made to it. Would be better to name it UtcCurrent or UtcTestTime.

**Comment by trwalke** [comment]
>         public DateTime GetUtcNow()

Do we need the method and the parameter?

---
Refers to: tests/Microsoft.Identity.Test.Common/Core/Mocks/TestTimeService.cs:23 in 23862eb. [](commit_id = 23862eb71ae0a0e619384bdfc514535489a51d8d, deletion_comment = False)

File: tests/Microsoft.Identity.Test.Unit/pop/CryptoProviderTests.cs [hunk 4]
```diff
@@ -5,7 +5,7 @@
 using Microsoft.Identity.Json.Linq;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 
-namespace Microsoft.Identity.Test.Unit.PoP
+namespace Microsoft.Identity.Test.Unit.pop
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Unit/pop/CryptoProviderTests.cs:8]: Should this be Pop instead?

**Comment by trwalke** [comment]
> using Microsoft.Identity.Client.AuthScheme.PoP;

Should this also be Pop for the source code?

---
Refers to: tests/Microsoft.Identity.Test.Unit/pop/CryptoProviderTests.cs:4 in 23862eb. [](commit_id = 23862eb71ae0a0e619384bdfc514535489a51d8d, deletion_comment = False)

**Review by trwalke — APPROVED** [review]
Makes sense to me.

**Comment by pmaytak** [comment]
@bgavrilMS Does this need to be rebased properly from the other pop branch?

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs [hunk 5]
```diff
@@ -529,6 +530,79 @@ public async Task CertificateOverrideAsync()
             }
         }
 
+        [TestMethod]
+        public async Task CertificateOverrideAsync()
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs:534]: This seems to be an old method, should be removed.
Also cert on line 482 above is not actually added to the app.

**Review by pmaytak — APPROVED** [review]

**Pull request closed without merging**